Do not use wxRegEx in wxCmpNaturalGeneric()

PBfordev opened the pull request:
Proposal based on the discussion here https://github.com/wxWidgets/wxWidgets/commit/371c4b13661641161f2c71ac2bc1044e48681705#commitcomment-41193189

TBH, I have not tested if `wxIsspace() `and `wxIspunct()` work properly with Unicode space and punctuation characters (besides checking `wxIsspace(L'\u2003') == true` on MSW), hoping that the CRT got it right...

PBfordev committed:
Do not use wxRegEx in wxCmpNaturalGeneric()

Using wxRegEx in wxCmpNaturalGeneric() introduced a dependency of the base library
on the regex library.

Replace wxRegEx with character classification functions wxIsspace(), wxIspunct(),
and wxIsdigit() to remove this rather unnecessary dependency.

PBfordev commented on src/common/arrstr.cpp:
```diff
@@ -759,47 +756,48 @@ struct wxStringFragment
+    for ( it = text.begin(); it != text.end(); ++it )
     {
-        if ( reDigit.GetMatch(NULL, &length) )
+        const wxChar           ch = *it;
```
`ch` should probably be `wxUniChar` instead of `wxChar`, to make this actually work in UTF-8 builds....

PBfordev committed:
Improve on previous commit

Use wxUniChar instead of wxChar to make things work in UTF-8 builds as well.

Use wxString::const_iterator instead of wxString::iterator.

Possibly use not 18 but 19 characters for a number, same as in the original regex-using code.

Fix whitespace.

vadz commented on src/common/arrstr.cpp:
```diff
@@ -759,47 +756,47 @@ struct wxStringFragment
+        // string fragment type than the previously processed characters had
+        // or when there is a sequence of digits too long to fit in wxUint64
+        if ( fragment.type != chType ||
+             (fragment.type == wxStringFragment::Digit && it - text.begin() == 20) )
```
Very minor, but could we please define a symbolic constant for this `20`?

PBfordev replied:
Done.

And as always: If possible, could you please squash all the commits into one if merged. Thanks.

vadz replied:
> Also perhaps @catalinr who reviewed the original natural sort PR may have something to say here.

Sure, I'll wait for a few days before merging this.
 
> And Vadim, as always: If possible, could you please squash all the commits into one if merged. Thanks.

Will do. BTW, I recommend using `git commit --fixup` (and/or `--squash`), this makes it perfectly clear which commits must be combined and even allows Git to do it on its own (with autosquash option which is set in my `.gitconfig`).

vadz reviewed (approved):
Thanks a lot, not only this solves the dependency problem you found and probably makes the code more efficient, but it also makes it much more clear IMO, which is very welcome.

PBfordev committed:
Implement changes requested in review

Use a constant instead of literal for the maximum digit sequence length.

Also make some whitespace changes.

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -759,47 +756,52 @@ struct wxStringFragment
-    size_t           length = 0;
+    // the maximum length of a sequence of digits that
+    // can fit into wxUint64 when converted to a number
+    static const size_t maxDigitSequenceLength = 19;
```
```suggestion
    static const ptrdiff_t maxDigitSequenceLength = 19;
```
should fix the build failures (although I didn't test it).

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -759,47 +756,52 @@ struct wxStringFragment
+    for ( it = text.begin(); it != text.end(); ++it )
     {
-        if ( reDigit.GetMatch(NULL, &length) )
+        const wxUniChar        ch = *it;
```
```suggestion
        const wxUniChar&       ch = *it;
```
probably with little impact, but in the end it should avoid another copy of each string.

catalinr commented on src/common/arrstr.cpp:
```diff
@@ -759,47 +756,52 @@ struct wxStringFragment
     {
-        if ( reDigit.GetMatch(NULL, &length) )
+        const wxUniChar        ch = *it;
+        wxStringFragment::Type chType = wxStringFragment::Empty;
```
```suggestion
        wxStringFragment::Type chType;
```
The initial value is not needed since the code below ensures that the variable will always have a value assigned (thus there should be no compiler warnings either). But disregards this if you think it affects readability.

PBfordev committed:
Further improve the PR

Fix build by using ptrdiff_t instead of size_t for iterator distance.

Use cbegin()/cend() instead of begin()/end().

Use const reference for a currently processed wxUniChar.

PBfordev commented:
@catalinr, thanks for the review. I have implemented all your suggestions except for not initializing `chType` which I rather keep that way even if it is not useful there.

vadz commented:
Thanks again for improving this!
